web3.js/src/util/send-and-confirm-transaction.ts

```diff
@@ -34,36 +40,64 @@ export async function sendAndConfirmTransaction(
     sendOptions,
   );
 
-  const status = await connection.getSignatureStatus(signature);
+  const subscriptionCommitment =
+    options?.preflightCommitment || options?.commitment;
+  let subscriptionId: number | undefined;
+  let status: RpcResponseAndContext<SignatureResult> | null = null;
+
+  const confirmTx = new Promise((resolve, reject) => {
+    try {
+      subscriptionId = connection.onSignature(
+        signature,
+        (result: SignatureResult, context: Context) => {
+          subscriptionId = undefined;
+          status = {
+            context,
+            value: result,
+          };
+          resolve(null);
+        },
+        subscriptionCommitment,
+      );
+    } catch (err) {
+      reject(err);
+    }
+  });
 
   const checkBlockHeight = async () => {
     const blockHeight = await connection.getBlockHeight(
       options && options.commitment,
     );
 
-    if (blockHeight > transaction.lastValidBlockHeight!) {
-      throw new Error('Transaction has expired.');
-    } else {
-      if (status) {
-        return;
+    return blockHeight;
+  };
+
+  let currentBlockHeight = await checkBlockHeight();
+
+  const expireTx = async (): Promise<void> => {
+    while (currentBlockHeight <= transaction.lastValidBlockHeight!) {
+      if (transaction.lastValidBlockHeight! - currentBlockHeight > 200) {
+        await sleep(2000);
+      } else if (transaction.lastValidBlockHeight! - currentBlockHeight > 100) {
+        await sleep(1000);
+      } else {
+        await sleep(500);
       }
-      await sleep(500);
-      await checkBlockHeight();
+      currentBlockHeight = await checkBlockHeight();
     }
   };
 
-  await checkBlockHeight();
-  // const status = (
-  //   await connection.confirmTransaction(
-  //     signature,
-  //     options && options.commitment,
-  //   )
-  // ).value;
+  try {
+    await promiseExpiration(confirmTx, expireTx);
```

Comment:
I tried to implement this, but it was tricky because `expireTx` uses `async/await`, so converting it to `Promise` syntax proved to be challenging. I think what I have is correct. I'd like your feedback on `expireTx` specifically.

Reply:
Right. There are a couple of ways around this. All of these are equivalent. You can see how you can translate between them.

```
const result = await new Promise(resolve => {
  getResult().then(resolve);
});
```

```
const result = await new Promise(async resolve => {
  const result = await getResult();
  resolve(result);
});
```

```
const result = await new Promise(resolve => {
  (async () => {
    const result = await getResult();
    resolve(result);
  })();
});
```

```
const result = await (async {
  return await getResult();
})();
```

In all four of these cases, you can read and write to a `let done` binding that lives _outside_ of them. This is how you can cancel your `while` loop if you know the transaction to have completed.